File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Easy/1002. Find Common Characters/Python solution/1002. Find Common Characters.py
class Solution:
    def commonChars(self, words: List[str]) -> List[str]:
        res = []

        # Initialize common_character_counts with the characters from the first word
        common_character_count = collections.Counter(words[0])

        # Store the characters of the current word in current_character_count
        for i in range(1, len(words)):
            current_character_count = collections.Counter(words[i])

            # Update common_character_count to keep the minimum count
            for ch in common_character_count.keys():
                common_character_count[ch] = min(common_character_count[ch], current_character_count[ch])

        # Append the common characters
        for ch, count in common_character_count.items():
            for _ in range(count):  # if count > 1, ch occured count times in all the words
                res.append(ch)
        
        return res
        # Or without the need for appending the common characters just return the following 
        # return common_character_counts.elements()

                      
            

File: A2SV-Community-Education/Div-3-Group-3-2024/LeetCode/Easy/1002. Find Common Characters/C++ solution/1002. Find Common Characters.cpp
class Solution {
public:
    vector<string> commonChars(vector<string>& words) {
        vector <string> result;
        // It's better to use an unordered map since the Time Complexity will be O(n) unlike the normal map which has O(nlog n) for sorting, to use it we muct include <unordered_map>
        unordered_map<char, int> common_character_count;

        // Initalize common_character_count with the first string in words
        for (char ch : words[0])
            common_character_count[ch]++;

        for (int i = 1; i < words.size(); i++)
        {
            // store the current character count starting form the second string in words
            unordered_map <char, int> current_character_count;
            for (char ch : words[i])
                current_character_count[ch]++;

            // Update common_character_count to keep the minimum, this step will keep only the common character
            for (auto& pair : common_character_count)
            {
                // To be able to use the min function we must inclued algorithm
                pair.second = min (pair.second, current_character_count[pair.first]);
            }
        }

        // Append the common characters to result with respect to thier frequencies
        for (auto& pair : common_character_count)
        {
            for (int i = 0; i < pair.second; i++)
                result.push_back(string(1, pair.first));  // Convert char to string since we cannot return a vector that contain characters
        }

        return result;
    }
};